### ISSUE TITLE
There should be a way to abort wasm compilation

### ISSUE THREAD

**Issue by chicoxyzzy** [WebAssembly/design#1262]
Compilation (and instantiation?) of WebAssembly module could be a resource intensive operation so there should be a way for wasm hosts to abort this step in cases when module is not needed anymore for some reasons.
As for JS hosts, theoretically cancellation methods should be in JS API, but JS doesn't have [cancellation mechanism](https://github.com/tc39/proposal-cancellation/issues/22) for promises yet. There are [AbortController and AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) which are used to abort Fetch. Probably they can be used to abort wasm compilation (passing signal as extra parameter to `WebAssembly.compile`), but they are part of DOM WHATWG spec, so it's only possible to use them in browsers and as a consequence only in Web API.

**Comment by xtuc** [comment]
Note that Promise and Fetch run on the main thread, whereas Wasm compilation (if asynchronous) can be done in background, only the instantiation is blocking.

**Comment by chicoxyzzy** [comment]
> whereas Wasm compilation (if asynchronous) can be done in background
Yes, but the problem is not main thread blocking, but unnecessary operations (fetch also doesn't block main thread). Also it's related to any host, not just browsers.

**Comment by chicoxyzzy** [comment]
> Also I'm wondering what happens when an module (while compiling) gets garbage collected.

Good question. I think we can test this case aborting fetch while `instantiateStreaming` / `compileStreaming`.

**Comment by xtuc** [comment]
> but any ongoing background compilation has a reference to the Module object and will keep it alive until the compilation is finished

Is there any plan to stop the compilation in that case? I don't know how common that situation will be.

**Comment by hammacher** [comment]
> > Also I'm wondering what happens when an module (while compiling) gets garbage collected.
> 
> In Firefox the JS WasmModule object is a front for an internal Module object which is shared among the threads that have a handle to the module; the object is (atomically) refcounted. When the JS WasmModule object is GC'd it just drops its refcount on the Module, but any ongoing background compilation has a reference to the Module object and will keep it alive until the compilation is finished.

V8 does it very similarly, except that compilation does not keep the module alive. We cancel compilation when the ref count drops to zero.
But note that the WasmModule is only created when baseline compilation finishes, so up until this point compilation cannot be cancelled because there is a promise to be resolved. So cancellation really only applies to tiering.

**Comment by chicoxyzzy** [comment]
@lukewagner this won't work for Node.js and other non-browser environments.

**Comment by lukewagner** [comment]
You're right, the streaming APIs are in the Web API, but I'm not sure if the solution is to have wasm's JS API to invent its own duplicate version of `AbortController`; ideally `fetch()`/`Response`/`AbortController` could be implemented in these non-browser environments, thereby enabling cross-environment portable, efficient streaming wasm compilation.  I don't know how politically loaded of a solution that is, though.

**Comment by chicoxyzzy** [comment]
Cancellation won't be possible at all until we'll have an API for that in `compile` and `instantiate` methods.

**Comment by chicoxyzzy** [comment]
We now have at least one precedent of using `AbortController` outside of DOM spec — [Web NFC now also uses it](https://github.com/w3c/web-nfc/blob/gh-pages/EXPLAINER.md#integration-with-abortcontroller).

**Comment by chicoxyzzy** [comment]
[Web Locks API also uses `AbortController`](https://github.com/WICG/web-locks/blob/master/EXPLAINER.md#signal-option-to-request). 

Related discussions:
- [Web NFC repo](https://github.com/w3c/web-nfc/issues/147)
- [W3C TAG design principles repo](https://github.com/w3ctag/design-principles/issues/138)
- some discussions about `AbortController` in Node.js repo
  * [Web Locks API](https://github.com/nodejs/node/issues/22702)
  * [Timers](https://github.com/nodejs/node/pull/29521) (+discussion on timers [in HTML repo](https://github.com/whatwg/html/issues/617))
